modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckProcess.java

```diff
@@ -0,0 +1,543 @@
+
+        return err;
+    }
+
+    /** */
+    private Map<ClusterNode, Exception> collectErrors(Map<UUID, Throwable> errors, Set<UUID> requiredNodes) {
+        if (errors.isEmpty())
+            return Collections.emptyMap();
+
+        return errors.entrySet().stream().filter(e -> requiredNodes.contains(e.getKey()) && e.getValue() != null)
+            .collect(Collectors.toMap(e -> kctx.cluster().get().node(e.getKey()), e -> asException(e.getValue())));
+    }
+
+    /** */
+    private Map<ClusterNode, Map<PartitionKeyV2, PartitionHashRecordV2>> collectPartsHashes(
+        @Nullable Map<UUID, ? extends Map<PartitionKeyV2, PartitionHashRecordV2>> results,
+        Collection<UUID> requiredNodes
+    ) {
+        if (results == null)
+            return Collections.emptyMap();
+
+        return results.entrySet().stream()
+            .filter(e -> requiredNodes.contains(e.getKey()) && e.getValue() != null)
+            .collect(Collectors.toMap(e -> kctx.cluster().get().node(e.getKey()), Map.Entry::getValue));
+    }
+
+    /**
+     * @param snpName Snapshot name of the validation process. If {@coe null}, ignored.
+     * @param procId  If {@code snpName} is {@code null}, is used to find the operation request.
+     * @return Current snapshot checking request by {@code snpName} or {@code procId}.
+     */
+    private @Nullable SnapshotCheckProcessRequest currentRequest(@Nullable String snpName, UUID procId) {
+        return snpName == null
+            ? requests.values().stream().filter(rq -> rq.requestId().equals(procId)).findFirst().orElse(null)
+            : requests.get(snpName);
+    }
+
+    /** Phase 1 beginning: prepare, collect and check local metas. */
+    private IgniteInternalFuture<ArrayList<SnapshotMetadata>> prepareAndCheckMetas(SnapshotCheckProcessRequest extReq) {
+        SnapshotCheckProcessRequest locReq = requests.computeIfAbsent(extReq.snapshotName(), snpName -> extReq);
+
+        if (!locReq.equals(extReq)) {
+            Throwable err = new IllegalStateException("Validation of snapshot '" + extReq.snapshotName()
+                + "' has already started. Request=" + locReq + '.');
+
+            clean(extReq.requestId(), err, null, null);
+
+            return new GridFinishedFuture<>(err);
+        }
+
+        if (!extReq.nodes.contains(kctx.localNodeId())) {
+            if (log.isDebugEnabled()) {
+                log.debug("Skipping snapshot local metadatas collecting for snapshot validation, request=" + extReq
+                    + ". Current node is not required.");
+            }
+
+            return new GridFinishedFuture<>();
+        }
+
+        GridFutureAdapter<ArrayList<SnapshotMetadata>> locMetasChkFut = new GridFutureAdapter<>();
+
+        assert locReq.fut() == null;
+
+        locReq.fut(locMetasChkFut);
+
+        IgniteSnapshotManager snpMgr = kctx.cache().context().snapshotMgr();
+
+        snpMgr.snapshotExecutorService().submit(() -> stopFutureOnAnyFailure(locMetasChkFut, () -> {
```

Comment:
As I can see in the code of `SnapshotChecker` it already uses the executor service for some checks. Let's delegate all work with executor services to the checker.